### PR TITLE
feat: support wildcards in routingkeys

### PR DESCRIPTION
**Pull request by peter-svensson**

### CODE REVIEW

File: internal/handlers/matcher.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+// overlaps checks if two AMQP binding patterns overlap
+func overlaps(p1, p2 string) bool {
+	if p1 == p2 {
+		return true
+	}
+	r1 := strings.ContainsAny(p1, "#*")
+	r2 := strings.ContainsAny(p2, "#*")
+
+	if r1 {
+		return match(p1, p2)
```
Comment (argoyle) [inline internal/handlers/matcher.go]: What if p1 doesn't match p2 but p2 matches p1? Not sure if that's even possible. That would mean that both are wildcards. p1: user.#.a and p2: user.# would be such a case, right?

File: connection.go [hunk 2]
```diff
@@ -54,12 +56,12 @@ type Route struct {
...
-	amqpUri     amqp.URI
-	connection  amqpConnection
-	channel     AmqpChannel
-	handlers    map[queueRoutingKey]messageHandlerInvoker
+	started       bool
+	serviceName   string
+	amqpUri       amqp.URI
+	connection    amqpConnection
+	channel       AmqpChannel
+	queueHandlers handlers.QueueHandlers[messageHandlerInvoker]
```
Comment (argoyle) [inline connection.go]: Will we ever have any other type to QueueHandlers or could we move that into the type itself?

File: internal/handlers/handlers.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+import (
+	"fmt"
+)
+
+// Handler is the handler for a certain routing key and queue
+type Handler any
+
+// Handlers holds the handlers for a certain queue
+type Handlers[T Handler] struct {
```
Comment (argoyle) [inline internal/handlers/handlers.go]: Couldn't this just be an alias for `map[string]*T`? You don't really need the struct here.

File: internal/handlers/handlers.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	if !ok {
+		QueueHandlers = &Handlers[T]{}
+		h.all[queueName] = QueueHandlers
+	}
+
+	if mappedRoutingKey, exists := QueueHandlers.exists(routingKey); exists {
+		return fmt.Errorf("routingkey %s overlaps %s for queue %s, consider using AddQueueNameSuffix", routingKey, mappedRoutingKey, queueName)
+
+	}
+	QueueHandlers.add(routingKey, handler)
```
Comment (argoyle) [inline internal/handlers/handlers.go]:
```suggestion
	queueHandlers, ok := h.all[queueName]
	if !ok {
		queueHandlers = &Handlers[T]{}
		h.all[queueName] = queueHandlers
	}

	if mappedRoutingKey, exists := queueHandlers.exists(routingKey); exists {
		return fmt.Errorf("routingkey %s overlaps %s for queue %s, consider using AddQueueNameSuffix", routingKey, mappedRoutingKey, queueName)

	}
	queueHandlers.add(routingKey, handler)
```

Comment (argoyle) [reply internal/handlers/handlers.go]: The naming had me confused for a while here. 😄

File: internal/handlers/handlers_test.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+			key:      "user.#",
+			existing: []string{"user.a.updated"},
+			wantErr: func(t assert.TestingT, err error, i ...interface{}) bool {
+				return assert.EqualError(t, err, "routingkey user.# overlaps user.a.updated for queue q, consider using AddQueueNameSuffix")
+			},
+		},
+		{
+			name:     "overlap other wildcard",
+			key:      "user.#",
+			existing: []string{"user.a.*"},
```
Comment (argoyle) [inline internal/handlers/handlers_test.go:110]: Add another test where key and existing is reversed to prove my point from another comment

File: internal/handlers/matcher.go (hunk 1 above)
Comment (argoyle) [reply internal/handlers/matcher.go]: I think we only should return here if a match is found and not use else-if

**Review by argoyle — CHANGES_REQUESTED** [review]
A few comments

File: connection.go (hunk 2 above)
Comment (peter-svensson) [reply connection.go]: `QueueHandlers` is defined in another package...we would have to move `messageHandlerInvoker` then as well..and `messageHandlerInvoker` contains `HandlerFunc` so we will get a circular dependency unless we move it to a separate package as well. Not a big problem of course.
And the short answer to the question is "No, we will never have any other type" 😊

File: internal/handlers/handlers.go (hunk 3 above)
Comment (peter-svensson) [reply internal/handlers/handlers.go]: Should work just fine, had some more stuff there before but cleaning up 🧹

File: internal/handlers/matcher.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	"strings"
+)
+
+// overlaps checks if two AMQP binding patterns overlap
+func overlaps(p1, p2 string) bool {
+	if p1 == p2 {
+		return true
+	}
+	r1 := strings.ContainsAny(p1, "#*")
+	r2 := strings.ContainsAny(p2, "#*")
```
Comment (argoyle) [inline internal/handlers/matcher.go]: Maybe those checks are totally irrelevant?

**Review by argoyle — DISMISSED** [review]
Approved with comment. 🚀

File: connection.go (hunk 2 above)
Comment (argoyle) [reply connection.go]: True. Not worth the effort right now then.

**Review by argoyle — APPROVED** [review]

**Pull request merged**